### ISSUE TITLE
Rule Modification Request: Don't trigger `unavailable_function` for a function that returns `Never`

### ISSUE THREAD

**Issue by mattrubin** [realm/SwiftLint#3286]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Rule modification request

1. Why should this rule be added?

   Currently, the `unavailable_function` rule warns about every function that contains a `fatalError` and does not also contain a `return` statement. In most cases, it is correct to mark these functions as unavailable, but in a few small cases, functions which always crash are _intended_ to be called.

   Apple's documentation for [the `Never` type](https://developer.apple.com/documentation/swift/never) says:
   > Use `Never` as the return type when declaring a closure, function, or method that unconditionally throws an error, traps, or otherwise does not terminate.

   A return type of `Never` is an unambiguous signal to the caller that the function is not expected to return. I believe it should also be sufficient signal to SwiftLint that the function's author is not following the "use `fatalError` to avoid having to implement a method required by a protocol" pattern, and that the function is intended to be called.

2. Provide several examples of what _would_ and _wouldn't_ trigger violations.

   Wouldn't trigger:
   ```swift
   func resetOnboardingStateAndCrash() -> Never {
       resetUserDefaults()
       // Crash the app to re-start the onboarding flow.
       fatalError("Onboarding re-start crash.")
   }
   ```

   Would (still) trigger:
   ```swift
   func resetOnboardingStateAndCrash() {
       resetUserDefaults()
       // Crash the app to re-start the onboarding flow.
       fatalError("Onboarding re-start crash.")
   }
   ```

3. Should the rule be configurable? If so what parameters should be configurable?

   Maybe.

   I believe the use of `Never` is a strong enough signal of developer intention that this wouldn't need to be configurable, but if this change caused the rule to not warn in cases where others felt it should, an `include_functions_returning_never` configuration boolean could be added.

4. Should the rule (change) be opt-in or enabled by default? Why?

   Personally, I think the change should be enabled by default.

   Arguments for setting the `include_functions_returning_never` configuration to `false` by default, or not having a configuration option at all:
   - As noted above, I believe a function returning `Never` is clearly intended to be called.
   - Since the `unavailable_function` rule itself is opt-in, using an opinionated default once the rule has been voluntarily enabled seems less controversial than doing do for an enabled-by-default rule.

   Arguments for setting the `include_functions_returning_never` configuration to `true` by default:
   - This would maintain the current behavior by default (for users who opt in to the `unavailable_function` rule).
   - This would err on the side of _possibly-excessive_ warnings, rather than erring on the side of _not_ warning about some cases that the developer might have wanted to be warned about.

**Comment by mattrubin** [comment]
Commenting to prevent the stalebot from closing the issue: this would still be a helpful change.

**Comment by jpsim** [comment]
Agreed, I've updated the stale bot to exempt rule requests from being automatically closed in 6c5a850c.
